pkg/meta/base.go

```diff
@@ -230,6 +246,90 @@ func (m *baseMeta) checkRoot(inode Ino) Ino {
 	}
 }
 
+func (m *baseMeta) countDirUsage(ctx Context, ino Ino) (space, inodes uint64, err syscall.Errno) {
+	var entries []*Entry
+	if err = m.en.doReaddir(ctx, ino, 1, &entries, -1); err != 0 {
+		return
+	}
+	for _, e := range entries {
+		if string(e.Name) == "." || string(e.Name) == ".." {
+			continue
+		}
+		inodes += 1
+		var newSpace uint64
+		if e.Attr.Typ == TypeFile {
+			newSpace = e.Attr.Length
+		}
+		space += uint64(align4K(newSpace))
+	}
+	return
+}
+
+func (m *baseMeta) GetDirUsage(ctx Context, inode Ino) (space, inodes uint64, err error) {
+	return m.en.doGetDirUsage(ctx, m.checkRoot(inode))
+}
+
+func (m *baseMeta) increDirUsage(ctx Context, ino Ino, space int64, inodes int64) {
+	for i := 0; i < 50; i++ {
+		select {
+		case m.dirUsageEvents <- &dirUsageEvent{ino, space, inodes}:
```

Comment:
Most likely it will succeed, so maybe it's better to first try the channel once, and spawn the goroutine only when the channel is full.

Reply:
Resolved

Reply:
We should aggregate these updates in memory (a map), don't use channel here.

Reply:
Done